slatedb/src/db_state.rs

```diff
@@ -486,52 +477,31 @@ impl DbState {
         Ok(())
     }
 
-    pub fn move_imm_memtable_to_l0(
-        &mut self,
-        imm_memtable: Arc<ImmutableMemtable>,
-        sst_handle: SsTableHandle,
-    ) -> Result<(), SlateDBError> {
-        let mut state = self.state_copy();
-        let popped = state
-            .imm_memtable
-            .pop_back()
-            .expect("expected imm memtable");
-        assert!(Arc::ptr_eq(&popped, &imm_memtable));
-        state.manifest.core.l0.push_front(sst_handle);
-        state.manifest.core.replay_after_wal_id = imm_memtable.recent_flushed_wal_id();
-
-        // ensure the persisted manifest tick never goes backwards in time
-        let memtable_tick = imm_memtable.table().last_tick();
-        state.manifest.core.last_l0_clock_tick =
-            cmp::max(state.manifest.core.last_l0_clock_tick, memtable_tick);
-        if state.manifest.core.last_l0_clock_tick != memtable_tick {
-            return Err(SlateDBError::InvalidClockTick {
-                last_tick: state.manifest.core.last_l0_clock_tick,
-                next_tick: memtable_tick,
-            });
-        }
-        // update the persisted manifest last_l0_seq as the latest seq in the imm.
-        if let Some(seq) = imm_memtable.table().last_seq() {
-            state.manifest.core.last_l0_seq = seq;
-        }
-
-        self.update_state(state);
-        Ok(())
+    pub fn merge_remote_manifest(&mut self, remote_manifest: DirtyManifest) {
+        self.modify(|modifier| modifier.merge_remote_manifest(remote_manifest));
     }
 
-    /// increment the next wal id, and return the previous value.
-    pub fn increment_next_wal_id(&mut self) -> u64 {
-        let mut state = self.state_copy();
-        let next_wal_id = state.manifest.core.next_wal_sst_id;
-        state.manifest.core.next_wal_sst_id += 1;
-        self.update_state(state);
-        next_wal_id
+    pub fn modify<F, R>(&mut self, fun: F) -> R
+    where
+        F: FnOnce(&mut StateModifier<'_>) -> R,
+    {
+        let mut modifier = StateModifier::new(self);
+        let result = fun(&mut modifier);
+        modifier.finish();
+        result
     }
+}
+
+pub(crate) struct StateModifier<'a> {
+    db_state: &'a mut DbState,
+    pub state: COWDbState,
+}
 
-    pub fn set_next_wal_id(&mut self, next_wal_id: u64) {
-        let mut state = self.state_copy();
-        state.manifest.core.next_wal_sst_id = next_wal_id;
-        self.update_state(state);
+impl<'a> StateModifier<'a> {
```

Comment:
we can add methods that are commonly used in here so that we can chain multiple updates into one copy-on-write block (e.g. I could `merge_remote_manifest` then keep making some changes within the same `modify` block)